ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Run.hs

```diff
@@ -286,13 +256,39 @@ runPointSchedule schedulerConfig GenesisTest {gtSecurityParam = k, gtBlockTree}
     -- peer fragments than registered clients.
     let getCandidates = traverse readTVar candidates
     PeerSimulator.BlockFetch.startBlockFetchLogic registry chainDb fetchClientRegistry getCandidates
-    runScheduler schedulerConfig tracer pointSchedule resources
-    svChainSyncExceptions <- collectExceptions (Map.elems chainSyncRess)
-    svSelectedChain <- atomically $ ChainDB.getCurrentChain chainDb
-    pure $ StateView {
-      svSelectedChain,
-      svChainSyncExceptions
-      }
+
+    -- Scheduler main loop. It keeps track of the point schedule ran so far and
+    -- of the point schedule to be ran. It keeps extending the point schedule
+    -- for as long as 'scExtender' tells it to. It returns a final state view.
+    let runSchedulerExtending :: Maybe PointSchedule -> PointSchedule -> m StateView
+        runSchedulerExtending pastPointSchedule pointSchedule@PointSchedule{ticks} = do
+          -- Run the given point schedule: iterate over its ticks, sending each
+          -- tick to the associated peer in turn, giving each peer a chunk of
+          -- computation time, sequentially, until it satisfies the conditions
+          -- given by the tick. This usually means for the ChainSync server to
+          -- have sent the target header/s to the client.
+          traceWith tracer "Schedule is:"
+          for_ ticks  $ \tick -> traceWith tracer $ "  " ++ condense tick
+          for_ ticks (dispatchTick schedulerConfig tracer resources)
+          -- Build a state view. This allows inspecting the run so far. This
+          -- will be passed to the point schedule extender or returned to the tests.
+          svChainSyncExceptions <- collectExceptions (Map.elems chainSyncRess)
+          svSelectedChain <- atomically $ ChainDB.getCurrentChain chainDb
+          let svPointSchedule = maybe pointSchedule (<> pointSchedule) pastPointSchedule
+              stateView = StateView {
+                svSelectedChain,
+                svChainSyncExceptions,
+                svPointSchedule
+                }
+          -- Extend and loop or return.
+          case scExtender schedulerConfig stateView of
```

Comment:
Do you have a use case in mind where the extender would continue in the presence of exceptions? otherwise we could do the exception check unconditionally here.

Reply:
I don't think that'd be right. One peer can raise an exception (eg. because of a timeout) while the other peers still function properly.